File: TS3-LogDataViewer/parseXML.cpp
// parseXML.cpp : Parsing of the XML.
// Author : Drumsticks1
// GitHub : https://github.com/Drumsticks1/TS3-LogDataViewer

#include <vector>
#include <string>
#include "Constants.h"
#include "ServerGroup.h"
#include "Client.h"
#include "Ban.h"
#include "Kick.h"
#include "Complaint.h"
#include "Upload.h"
#include "customStreams.h"
#include <boost/filesystem.hpp>
#include "src/pugixml.hpp"
#define PUGIXML_HEADER_ONLY
#include "src/pugixml.cpp"

std::vector <std::string> parsedLogs;
extern std::vector <ServerGroup> ServerGroupList;
extern std::vector <Client> ClientList;
extern std::vector <Ban> BanList;
extern std::vector <Kick> KickList;
extern std::vector <Complaint> ComplaintList;
extern std::vector <Upload> UploadList;
extern unsigned int VIRTUALSERVER;
extern TeeStream outputStream;

// Parses the XML if existing.
bool parseXML() {
	if (boost::filesystem::exists(XMLFILE)) {
		if (boost::filesystem::is_regular_file(XMLFILE)) {
			if (!boost::filesystem::is_empty(XMLFILE)) {
				unsigned int ID, BanListID = 0, KickListID = 0, ComplaintListID = 0, UploadListID = 0, bannedID, bannedByID, bantime, kickedID, complaintAboutID, complaintByID, channelID, uploadedByID, deletedByID;
				std::string banDateTime, bannedNickname, bannedIP, bannedUID, bannedByNickname, bannedByUID, banReason,
					kickDateTime, kickedNickname, kickedByNickname, kickedByUID, kickReason,
					complaintDateTime, complaintAboutNickname, complaintReason, complaintByNickname,
					uploadDateTime, filename, uploadedByNickname, deletedByNickname;
				pugi::xml_document oldXML;

				outputStream << "Parsing old XML..." << std::endl;

				pugi::xml_parse_result result = oldXML.load_file(XMLFILE);
				if (!result) {
					outputStream << "An error occured while parsing the XML:" << std::endl << result.description() << std::endl;
					return false;
				}

				try {
					for (pugi::xml_node AttributesNode = oldXML.first_child().child("Attributes"); AttributesNode; AttributesNode = AttributesNode.next_sibling("Attributes")) {
						if (AttributesNode.child("VirtualServer").first_child().value() != std::to_string(VIRTUALSERVER)) {
							outputStream << "The last XML was created for another virtual server - skipping use of XML..." << std::endl;
							return false;
						}

						for (pugi::xml_node ParsedLogs = AttributesNode.child("ParsedLogs").child("P"); ParsedLogs; ParsedLogs = ParsedLogs.next_sibling("P")) {
							parsedLogs.emplace_back(ParsedLogs.first_child().value());
						}
					}

					for (pugi::xml_node ServerGroupNode = oldXML.first_child().child("ServerGroup"); ServerGroupNode; ServerGroupNode = ServerGroupNode.next_sibling("ServerGroup")) {
						if ((std::string)ServerGroupNode.child("ID").first_child().value() != "-1") {
							ID = std::stoul(ServerGroupNode.child("ID").first_child().value());
							ServerGroupList.resize(ID + 1);
							ServerGroupList[ID].addServerGroupInformation(ID, ServerGroupNode.child("ServerGroupName").first_child().value(),
								ServerGroupNode.child("CreationDateTime").first_child().value());

							if ((std::string)ServerGroupNode.child("Deleted").first_child().value() == "1") {
								ServerGroupList[ID].deleteServerGroup();
							}
						}
					}

					for (pugi::xml_node ClientNode = oldXML.first_child().child("Client"); ClientNode; ClientNode = ClientNode.next_sibling("Client")) {
						if ((std::string)ClientNode.child("ID").first_child().value() != "-1") {
							ID = std::stoul(ClientNode.child("ID").first_child().value());
							ClientList.resize(ID + 1);
							ClientList[ID].addID(ID);

							for (pugi::xml_node Nicknames = ClientNode.child("Nicknames").first_child(); Nicknames; Nicknames = Nicknames.next_sibling()) {
								ClientList[ID].addNicknameReverse(Nicknames.first_child().value());
							}

							for (pugi::xml_node Connections = ClientNode.child("Connections").first_child(); Connections; Connections = Connections.next_sibling()) {
								ClientList[ID].addDateTimeReverse(Connections.first_child().value());
							}

							for (pugi::xml_node IPs = ClientNode.child("IPs").first_child(); IPs; IPs = IPs.next_sibling()) {
								ClientList[ID].addIPReverse(IPs.first_child().value());
							}

							if ((std::string)ClientNode.child("Deleted").first_child().value() == "1") {
								ClientList[ID].deleteClient();
							}

							std::vector <unsigned int> bufferServerGroupIDs;
							std::vector <std::string> bufferServerGroupAssignmentDateTimes;

							for (pugi::xml_node ServerGroupIDs = ClientNode.child("ServerGroupIDs").first_child(); ServerGroupIDs; ServerGroupIDs = ServerGroupIDs.next_sibling()) {
								bufferServerGroupIDs.push_back(std::stoul(ServerGroupIDs.first_child().value()));
							}

							for (pugi::xml_node ServerGroupAssignmentDateTimes = ClientNode.child("ServerGroupAssignmentDateTimes").first_child();
							ServerGroupAssignmentDateTimes; ServerGroupAssignmentDateTimes = ServerGroupAssignmentDateTimes.next_sibling()) {
								bufferServerGroupAssignmentDateTimes.push_back(ServerGroupAssignmentDateTimes.first_child().value());
							}

							for (unsigned int i = 0; i < bufferServerGroupIDs.size(); i++)
								ClientList[ID].addServerGroup(bufferServerGroupIDs[i], bufferServerGroupAssignmentDateTimes[i]);
						}
					}

					for (pugi::xml_node BanNode = oldXML.first_child().child("Ban"); BanNode; BanNode = BanNode.next_sibling("Ban")) {
						banDateTime = BanNode.child("DateTime").first_child().value();
						bannedID = std::stoul(BanNode.child("ID").first_child().value());
						bannedNickname = BanNode.child("Nickname").first_child().value();
						bannedUID = BanNode.child("UID").first_child().value();
						bannedIP = BanNode.child("IP").first_child().value();
						bannedByNickname = BanNode.child("ByNickname").first_child().value();
						bannedByID = std::stoul(BanNode.child("ByID").first_child().value());
						bannedByUID = BanNode.child("ByUID").first_child().value();
						banReason = BanNode.child("Reason").first_child().value();
						bantime = std::stoul(BanNode.child("Bantime").first_child().value());

						BanList.resize(BanListID + 1);
						BanList[BanListID].addBan(banDateTime, bannedID, bannedNickname, bannedUID, bannedIP, bannedByNickname, bannedByID, bannedByUID, banReason, bantime);
						BanListID++;
					}

					for (pugi::xml_node KickNode = oldXML.first_child().child("Kick"); KickNode; KickNode = KickNode.next_sibling("Kick")) {
						kickDateTime = KickNode.child("DateTime").first_child().value();
						kickedID = std::stoul(KickNode.child("ID").first_child().value());
						kickedNickname = KickNode.child("Nickname").first_child().value();
						kickedByNickname = KickNode.child("ByNickname").first_child().value();
						kickedByUID = KickNode.child("ByUID").first_child().value();
						kickReason = KickNode.child("Reason").first_child().value();

						KickList.resize(KickListID + 1);
						KickList[KickListID].addKick(kickDateTime, kickedID, kickedNickname, kickedByNickname, kickedByUID, kickReason);
						KickListID++;
					}

					for (pugi::xml_node ComplaintNode = oldXML.first_child().child("Complaint"); ComplaintNode; ComplaintNode = ComplaintNode.next_sibling("Complaint")) {
						complaintDateTime = ComplaintNode.child("DateTime").first_child().value();
						complaintAboutNickname = ComplaintNode.child("AboutNickname").first_child().value();
						complaintAboutID = std::stoul(ComplaintNode.child("AboutID").first_child().value());
						complaintReason = ComplaintNode.child("Reason").first_child().value();
						complaintByNickname = ComplaintNode.child("ByNickname").first_child().value();
						complaintByID = std::stoul(ComplaintNode.child("ByID").first_child().value());

						ComplaintList.resize(ComplaintListID + 1);
						ComplaintList[ComplaintListID].addComplaint(complaintDateTime, complaintAboutNickname, complaintAboutID, complaintReason, complaintByNickname, complaintByID);
						ComplaintListID++;
					}

					for (pugi::xml_node UploadNode = oldXML.first_child().child("Upload"); UploadNode; UploadNode = UploadNode.next_sibling("Upload")) {
						uploadDateTime = UploadNode.child("DateTime").first_child().value();
						channelID = std::stoul(UploadNode.child("ChannelID").first_child().value());
						filename = UploadNode.child("Filename").first_child().value();
						uploadedByNickname = UploadNode.child("UplByNickname").first_child().value();
						uploadedByID = std::stoul(UploadNode.child("UplByID").first_child().value());

						UploadList.resize(UploadListID + 1);

						if ((std::string)UploadNode.child("DelByID").first_child().value() != "") {
							deletedByNickname = UploadNode.child("DelByNickname").first_child().value();
							deletedByID = std::stoul(UploadNode.child("DelByID").first_child().value());
							UploadList[UploadListID].addUpload(uploadDateTime, channelID, filename, uploadedByNickname, uploadedByID, deletedByNickname, deletedByID);
						}
						else UploadList[UploadListID].addUpload(uploadDateTime, channelID, filename, uploadedByNickname, uploadedByID);

						UploadListID++;
					}
					return true;
				}
				catch (std::exception& ex) {
					outputStream << "An error occured while parsing the XML:" << std::endl << ex.what() << std::endl;
					ServerGroupList.clear();
					ClientList.clear();
					BanList.clear();
					KickList.clear();
					ComplaintList.clear();
					UploadList.clear();
					return false;
				}
			}
			else return false;
		}
		else return false;
	}
	else return false;
}

File: nodejs/parseXML.js
// parseXML.js : Parsing of the XML.
// Author : Drumsticks1
// GitHub : https://github.com/Drumsticks1/TS3-LogDataViewer

"use strict";

// Todo: switch to a faster library!

const fs = require('fs'),
    main = require("./main.js"),
    Constants = require("./Constants.js"),
    globalArrays = require("./globalArrays.js"),
    miscFunctions = require("./miscFunctions.js"),
    outputHandler = require("./outputHandler.js"),
    xmldoc = require('xmldoc');

var parsedLogs = globalArrays.parsedLogs,
    ClientList = globalArrays.ClientList,
    ServerGroupList = globalArrays.ServerGroupList,
    BanList = globalArrays.BanList,
    KickList = globalArrays.KickList,
    ComplaintList = globalArrays.ComplaintList,
    UploadList = globalArrays.UploadList;

exports.parseXML = function() {
    try {
        var XMLStats = fs.statSync(Constants.XMLFILE);
        if (XMLStats.isFile()) {
            outputHandler.output("Parsing old XML...");

            try {
                var x = new Date();
                var XML = new xmldoc.XmlDocument(fs.readFileSync(Constants.XMLFILE));
                miscFunctions.updateCurrentDate();
                console.log(miscFunctions.getProgramRuntime(x));

                var ID, BanListID = 0, KickListID = 0, ComplaintListID = 0, UploadListID = 0, bannedID, bannedByID, bantime, kickedID, complaintAboutID, complaintByID, channelID, uploadedByID, deletedByID,
                    banDateTime, bannedNickname, bannedIP, bannedUID, bannedByNickname, bannedByUID, banReason,
                    kickDateTime, kickedNickname, kickedByNickname, kickedByUID, kickReason,
                    complaintDateTime, complaintAboutNickname, complaintReason, complaintByNickname,
                    uploadDateTime, filename, uploadedByNickname, deletedByNickname;

                var AttributesNode = XML.childNamed("Attributes");

                if (AttributesNode.childNamed("VirtualServer").val != String(main.VIRTUALSERVER)) {
                    outputHandler.output("The last XML was created for another virtual server - skipping use of XML...");
                    return false;
                }

                var ParsedLogNodes = XML.childNamed("ParsedLogs").children;
                for (var i = 0; i < ParsedLogNodes.length; i++) {
                    parsedLogs.push(ParsedLogNodes[i].val);
                }

                var ClientNodes = XML.childrenNamed("Client");
                for (i = 0; i < ClientNodes.length; i++) {
                    if (ClientNodes[i].childNamed("ID").val != "-1") {
                        ID = Number(ClientNodes[i].childNamed("ID").val);
                        ClientList.resizeFill(ID + 1, "Client");
                        ClientList[ID].addID(ID);

                        var Nicknames = ClientNodes[i].childrenNamed("Nicknames")[0].children;
                        for (var j = 0; j < Nicknames.length; j++) {
                            ClientList[ID].addNicknameReverse(Nicknames[j].val);
                        }

                        var Connections = ClientNodes[i].childrenNamed("Connections")[0].children;
                        for (j = 0; j < Connections.length; j++) {
                            ClientList[ID].addDateTimeReverse(Connections[j].val);
                        }

                        var IPs = ClientNodes[i].childrenNamed("IPs")[0].children;
                        for (j = 0; j < IPs.length; j++) {
                            ClientList[ID].addIPReverse(IPs[j].val);
                        }

                        if (ClientNodes[i].childNamed("Deleted") != undefined) {
                            ClientList[ID].deleteClient();
                        }

                        var bufferServerGroupIDs = [],
                            bufferServerGroupAssignmentDateTimes = [];

                        var ServerGroupIDs = ClientNodes[i].childrenNamed("ServerGroupIDs")[0].children;
                        for (j = 0; j < ServerGroupIDs.length; j++) {
                            bufferServerGroupIDs.push(Number(ServerGroupIDs[j].val));
                        }

                        var ServerGroupAssignmentDateTimes = ClientNodes[i].childrenNamed("ServerGroupAssignmentDateTimes")[0].children;
                        for (j = 0; j < ServerGroupAssignmentDateTimes.length; j++) {
                            bufferServerGroupAssignmentDateTimes.push(ServerGroupAssignmentDateTimes[j].val);
                        }

                        for (j = 0; j < bufferServerGroupIDs.length; j++) {
                            ClientList[ID].addServerGroup(bufferServerGroupIDs[j], bufferServerGroupAssignmentDateTimes[j]);
                        }
                    }
                }

                var ServerGroupNodes = XML.childrenNamed("ServerGroup");
                for (i = 0; i < ServerGroupNodes.length; i++) {
                    if (ServerGroupNodes[i].childNamed("ID").val != "-1") {
                        ID = Number(ServerGroupNodes[i].childNamed("ID").val);
                        ServerGroupList.resizeFill(ID + 1, "ServerGroup");
                        ServerGroupList[ID].addServerGroupInformation(ID, ServerGroupNodes[i].childNamed("ServerGroupName").val,
                            ServerGroupNodes[i].childNamed("CreationDateTime").val);

                        if (ServerGroupNodes[i].childNamed("Deleted") != undefined) {
                            ServerGroupList[ID].deleteServerGroup();
                        }
                    }
                }

                var BanNodes = XML.childrenNamed("Ban");
                for (i = 0; i < BanNodes.length; i++) {
                    banDateTime = BanNodes[i].childNamed("DateTime").val;
                    bannedID = Number(BanNodes[i].childNamed("ID").val);
                    bannedNickname = BanNodes[i].childNamed("Nickname").val;
                    bannedUID = BanNodes[i].childNamed("UID").val;
                    bannedIP = BanNodes[i].childNamed("IP").val;
                    bannedByNickname = BanNodes[i].childNamed("ByNickname").val;
                    bannedByID = Number(BanNodes[i].childNamed("ByID").val);
                    bannedByUID = BanNodes[i].childNamed("ByUID").val;
                    banReason = BanNodes[i].childNamed("Reason").val;
                    bantime = Number(BanNodes[i].childNamed("Bantime").val);

                    BanList.resizeFill(BanListID + 1, "Ban");
                    BanList[BanListID].addBan(banDateTime, bannedID, bannedNickname, bannedUID, bannedIP, bannedByNickname, bannedByID, bannedByUID, banReason, bantime);
                    BanListID++;
                }
                var KickNodes = XML.childrenNamed("Kick");
                for (i = 0; i < KickNodes.length; i++) {
                    kickDateTime = KickNodes[i].childNamed("DateTime").val;
                    kickedID = Number(KickNodes[i].childNamed("ID").val);
                    kickedNickname = KickNodes[i].childNamed("Nickname").val;
                    kickedByNickname = KickNodes[i].childNamed("ByNickname").val;
                    kickedByUID = KickNodes[i].childNamed("ByUID").val;
                    kickReason = KickNodes[i].childNamed("Reason").val;

                    KickList.resizeFill(KickListID + 1, "Kick");
                    KickList[KickListID].addKick(kickDateTime, kickedID, kickedNickname, kickedByNickname, kickedByUID, kickReason);
                    KickListID++;
                }

                var ComplaintNodes = XML.childrenNamed("Complaint");
                for (i = 0; i < ComplaintNodes.length; i++) {
                    complaintDateTime = ComplaintNodes[i].childNamed("DateTime").val;
                    complaintAboutNickname = ComplaintNodes[i].childNamed("AboutNickname").val;
                    complaintAboutID = Number(ComplaintNodes[i].childNamed("AboutID").val);
                    complaintReason = ComplaintNodes[i].childNamed("Reason").val;
                    complaintByNickname = ComplaintNodes[i].childNamed("ByNickname").val;
                    complaintByID = Number(ComplaintNodes[i].childNamed("ByID").val);

                    ComplaintList.resizeFill(ComplaintListID + 1, "Complaint");
                    ComplaintList[ComplaintListID].addComplaint(complaintDateTime, complaintAboutNickname, complaintAboutID, complaintReason, complaintByNickname, complaintByID);
                    ComplaintListID++;
                }

                var UploadNodes = XML.childrenNamed("Upload");
                for (i = 0; i < UploadNodes.length; i++) {
                    uploadDateTime = UploadNodes[i].childNamed("DateTime").val;
                    channelID = Number(UploadNodes[i].childNamed("ChannelID").val);
                    filename = UploadNodes[i].childNamed("Filename").val;
                    uploadedByNickname = UploadNodes[i].childNamed("UplByNickname").val;
                    uploadedByID = Number(UploadNodes[i].childNamed("UplByID").val);

                    UploadList.resizeFill(UploadListID + 1, "Upload");

                    if (UploadNodes[i].childNamed("DelByID") != undefined) {
                        deletedByNickname = UploadNodes[i].childNamed("DelByNickname").val;
                        deletedByID = Number(UploadNodes[i].childNamed("DelByID").val);
                        UploadList[UploadListID].addUpload(uploadDateTime, channelID, filename, uploadedByNickname, uploadedByID, deletedByNickname, deletedByID);
                    }
                    else UploadList[UploadListID].addUpload(uploadDateTime, channelID, filename, uploadedByNickname, uploadedByID);
                    UploadListID++;
                }

                return true;
            } catch (error) {
                outputHandler.output("An error occurred while parsing the XML:\n" + error.message);
                ClientList = [];
                ServerGroupList = [];
                BanList = [];
                KickList = [];
                ComplaintList = [];
                UploadList = [];
                return false;
            }

        } else {
            outputHandler.output(Constants.XMLFILE + " seems not to be a file...");
            return false;
        }
    }
    catch (error) {
        outputHandler.output("An error occurred while fetching the XML:\n" + (error.message));
        return false;
    }
};